modules: openthread: use PSA Crypto API as default choice for crypto

valeriosetti opened the pull request:
Modify OpenThread's Kconfig options in order to always use PSA Crypto API by default instead of legacy Mbed TLS crypto.
Please check commits messages for details.

These changes were tested with `zephyr/samples/net/openthread/coap/` on 2x `nrf52840dk` boards .

valeriosetti committed:
modules: openthread: move definition of OPENTHREAD_CRYPTO_PSA

Simply move the definition of OPENTHREAD_CRYPTO_PSA Kconfig closer to
the OPENTHREAD_MBEDTLS one. This will help in following commits.

Signed-off-by: Valerio Setti <vsetti@baylibre.com>

valeriosetti committed:
modules: openthread: add couple of internal helper Kconfigs

Add CONFIG_OPENTHREAD_COMMISSIONER_JOINER and
CONFIG_OPENTHREAD_SRP_CLI_SRV as internal helpers to slightly simplify
depenencies in other Kconfigs that depends on these combinations.

Signed-off-by: Valerio Setti <vsetti@baylibre.com>

valeriosetti committed:
modules: openthread: minor documentation fixes

Minor fixes to some Kconfigs' prompt and help messages.

Signed-off-by: Valerio Setti <vsetti@baylibre.com>

valeriosetti committed:
modules: openthread: rename security related Kconfigs

Rename
- CUSTOM_OPENTHREAD_SECURITY as OPENTHREAD_SECURITY_CUSTOM
- OPENTHREAD_MBEDTLS_CHOICE as OPENTHREAD_SECURITY_DEFAULT_CONFIG

Signed-off-by: Valerio Setti <vsetti@baylibre.com>

valeriosetti committed:
modules: openthread: reorganize default configuration options

First separate Mbed TLS' crypto configuration from the "high level" (TLS)
part. CONFIG_OPENTHREAD_MBEDTLS_TLS_CONFIG now only takes care of this high
level part, whereas the newly added
CONFIG_OPENTHREAD_CRYPTO_LEGACY_MBEDTLS_CONFIG takes care of the legacy
crypto configuration.

CONFIG_OPENTHREAD_CRYPTO_PSA_CONFIG is added to propose a default crypto
configuration based on PSA Crypto API instead of legacy Mbed TLS.

Both CONFIG_OPENTHREAD_CRYPTO_LEGACY_MBEDTLS and
CONFIG_OPENTHREAD_CRYPTO_PSA_CONFIG are promptless and the choice between
the two is based on CONFIG_OPENTHREAD_CRYPTO_PSA.

Signed-off-by: Valerio Setti <vsetti@baylibre.com>

valeriosetti committed:
modules: openthread: use PSA API by default

Make CONFIG_OPENTHREAD_CRYPTO_PSA as enabled by default in order to use
PSA Crypto API instead of legacy Mbed TLS' crypto. This is done with the
final goal to deprecate and remove usage of legacy crypto support in
the near future from Zephyr codebase.

Signed-off-by: Valerio Setti <vsetti@baylibre.com>

valeriosetti committed:
secure_storage: add dependency for Settings backend

Secure Storage is used as PSA ITS in builds where TF-M is not enabled,
but it needs a storage backend where to physically save data.
If CONFIG_SECURE_STORAGE_ITS_STORE_IMPLEMENTATION_ZMS is used then its
"depends on" guarantee that there is a partition to use as physical
storage.
If CONFIG_SECURE_STORAGE_ITS_STORE_IMPLEMENTATION_SETTINGS is used instead
we want to ensure that CONFIG_SETTINGS_NONE doesn't gets selected which
would result in Settings not being able to store data anywhere.

This commit add a depends_on for this last case.

Signed-off-by: Valerio Setti <vsetti@baylibre.com>

rlubos commented on doc/releases/migration-guide-4.3.rst:
```diff
@@ -516,6 +516,15 @@ LED Strip
 
 * Renamed ``arduino,modulino-smartleds`` to :dtcompatible:`arduino,modulino-pixels`
 
+OpenThread
```
This will likely not get to 4.3 given we're already in the freeze period, perhaps move it to `migration-guide-4.4.rst` already?

tomi-font replied:
^

rlubos commented on samples/net/openthread/coap/prj.conf:
```diff
@@ -26,3 +26,7 @@ CONFIG_OPENTHREAD_XPANID="11:11:11:11:22:22:22:22"
 CONFIG_OPENTHREAD_NETWORK_NAME="OpenThreadDemo"
+
+# Use ZMS as storage backend for Secure Storage
+CONFIG_SECURE_STORAGE_ITS_STORE_IMPLEMENTATION_ZMS=y
```
Wouldn't we need this in other samples as well? And those generic networking samples that have `overlay-ot.conf`?

rlubos replied:
This is still no clear to me, do we need thsoe configs in other samples/OT overlays as well?

valeriosetti replied:
I added the DNM label to ensure that this PR is not merged accidentally before we're happy with test coverage

valeriosetti commented on doc/releases/migration-guide-4.3.rst:
```diff
@@ -516,6 +516,15 @@ LED Strip
+  - ``CONFIG_OPENTHREAD_MBEDTLS_CHOICE`` to
+    :kconfig:option:`CONFIG_OPENTHREAD_SECURITY_DEFAULT_CONFIG`
+  - ``CONFIG_CUSTOM_OPENTHREAD_SECURITY`` to
+    :kconfig:option:`CONFIG_OPENTHREAD_SECURITY_CUSTOM`
```
> A separate question whether we need the duplicate _CONFIG at the end in both configs, it looks a bit odd.

Well, the first `CONFIG_` is for the Kconfig thing whereas the latter refers to the fact that this Kconfig relates to an Mbed TLS' configuration. True that are both abbreviated with `CONFIG`, but they refer to different things.
If you have any better wording I'm totally open to proposals :)

rlubos replied:
If the second config is justified then fine, not a big deal :+1:

tomi-font replied:
still missing `_CONFIG` suffix

tomi-font commented on modules/openthread/Kconfig:
```diff
@@ -179,22 +179,22 @@ config OPENTHREAD_MBEDTLS
+	select MBEDTLS_ECP_NIST_OPTIM if OPENTHREAD_COMMISSIONER_JOINER
+	select MBEDTLS_CTR_DRBG_C if OPENTHREAD_COMMISSIONER_JOINER
+	select MBEDTLS_HMAC_DRBG_C if OPENTHREAD_COMMISSIONER_JOINER || \
+					    OPENTHREAD_SRP_CLI_SRV
```
alignment's off

tomi-font commented on modules/openthread/Kconfig:
```diff
@@ -162,43 +162,87 @@ config OPENTHREAD_SECURITY_CUSTOM_CONFIG
 	bool "PSA Crypto API"
-	depends on PSA_CRYPTO_CLIENT
+	select PSA_CRYPTO
+	select SECURE_STORAGE if !BUILD_WITH_TFM
```
Is PSA Secure Storage actually a hard dependency? Why wasn't it enabled before?

tomi-font commented on doc/releases/migration-guide-4.3.rst:
```diff
@@ -525,6 +525,11 @@ OpenThread
   - ``CONFIG_CUSTOM_OPENTHREAD_SECURITY`` to
     :kconfig:option:`CONFIG_OPENTHREAD_SECURITY_CUSTOM`
 
+* In builds without TF-M :kconfig:option:`CONFIG_OPENTHREAD_CRYPTO_PSA` selects
```
```suggestion
* In builds without TF-M, :kconfig:option:`CONFIG_OPENTHREAD_CRYPTO_PSA` now selects

tomi-font commented on modules/openthread/Kconfig:
```diff
@@ -162,43 +162,87 @@ config OPENTHREAD_SECURITY_CUSTOM_CONFIG
+
+config OPENTHREAD_LEGACY_CRYPTO_CONFIG
+	bool
+	depends on !OPENTHREAD_CRYPTO_PSA
```
We could have a `choice` to select between legacy and PSA?

tomi-font commented on modules/openthread/CMakeLists.txt:
```diff
@@ -1,6 +1,12 @@
 
+if(CONFIG_OPENTHREAD_CRYPTO_PSA AND NOT CONFIG_BUILD_WITH_TFM)
+  if(CONFIG_SECURE_STORAGE_ITS_STORE_IMPLEMENTATION_SETTINGS AND CONFIG_SETTINGS_NONE)
+    message(FATAL_ERROR "No backend selected for Secure Storage.")
```
No **Settings** backend, rather?

tomi-font replied:
superseded by another suggestion

tomi-font commented on modules/openthread/Kconfig:
```diff
@@ -241,13 +241,17 @@ endif # OPENTHREAD_SECURITY_DEFAULT_CONFIG
 	  instead of the default, using Mbed TLS.
+	  Warning: disabling this Kconfig (enabled by default) and falling back
+	  to legacy Mbed TLS crypto API is deprecated and will be removed in the
+	  future.
```
It's indirectly selectable (enabled/disabled) through `OPENTHREAD_CRYPTO_PSA`, so to me it makes sense.

tomi-font commented on doc/releases/migration-guide-4.4.rst:
```diff
@@ -44,5 +44,22 @@ Other subsystems
+OpenThread
+==========
+
+* Kconfigs options renaming:
```
Maybe rather either
```suggestion
* Kconfigs options renamed:
```
or

```suggestion
* The following Kconfigs options were renamed:
```
?

valeriosetti committed:
samples: net: openthread: coap: enable NVS as Settings backend

When PSA Crypto API are used they need a PSA ITS implementation and this
is provided by Secure Storage if TF-M is not enabled in the build.
Secure Storage can use Settings as backend and Settings, on its turn, can
use several other backends to store data (ex: ZMS, NVS, etc).
Since NVS is already enabled in other OpenThread Zephyr samples, let's
enable NVS also for this sample.

While at it, also slightly re-organize "sample.yaml" to share more data
between tests and allow building and testing also on the nrf52840dk board.

Signed-off-by: Valerio Setti <vsetti@baylibre.com>

valeriosetti committed:
modules: mbedtls: allow ECJPAKE keyexchange to work through PSA API

Allow MBEDTLS_KEY_EXCHANGE_ECJPAKE_ENABLED to work with either legacy
crypto (i.e. MBEDTLS_ECJPAKE_C) or PSA Crypto API
(i.e. PSA_WANT_ALG_JPAKE).

Signed-off-by: Valerio Setti <vsetti@baylibre.com>

valeriosetti committed:
modules: openthread: select Secure Storage when using PSA without TF-M

OpenThread needs a PSA ITS provider when using PSA Crypto API, therefore
we enable Secure Storage in builds without TF-M.

Signed-off-by: Valerio Setti <vsetti@baylibre.com>

valeriosetti committed:
modules: openthread: add HKDF support in crypto_psa

crypto_psa.c was missing an implementation for the following functions:
- otPlatCryptoHkdfInit
- otPlatCryptoHkdfExtract
- otPlatCryptoHkdfExpand
- otPlatCryptoHkdfDeinit
which provide support for HKDF using PSA Crypto API.

This commit fill this gap by copying the implementation from upstream
PR 11445.

Signed-off-by: Valerio Setti <vsetti@baylibre.com>

valeriosetti committed:
manifest: openthread: bump version to get support for HKDF crypto alg

Bump OpenThread version to get support for HKDF crypto algorithm.

Signed-off-by: Valerio Setti <vsetti@baylibre.com>

tomi-font commented on modules/openthread/Kconfig:
```diff
@@ -162,48 +162,81 @@ config OPENTHREAD_SECURITY_CUSTOM_CONFIG
+
+endchoice # OPENTHREAD_CRYPTO_IMPLEMENTATION
+
+config OPENTHREAD_MBEDTLS_CONFIG
```
`MBEDTLS_TLS_CONFIG`? :upside_down_face:

tomi-font commented on modules/openthread/Kconfig:
```diff
@@ -176,6 +176,7 @@ choice OPENTHREAD_CRYPTO_IMPLEMENTATION
 config OPENTHREAD_CRYPTO_PSA
 	bool "PSA Crypto API"
 	select PSA_CRYPTO
+	select SECURE_STORAGE if !BUILD_WITH_TFM
```
Side note: maybe at some point we could have a `PSA_SECURE_STORAGE` Kconfig option that does the same as `PSA_CRYPTO` but for PSA Secure Storage.

tomi-font reviewed (commented):
mostly LGTM

Damian-Nordic commented on modules/openthread/Kconfig:
```diff
@@ -162,48 +162,81 @@ config OPENTHREAD_SECURITY_CUSTOM_CONFIG
-	select MBEDTLS_ENABLE_HEAP
+if OPENTHREAD_SECURITY_DEFAULT_CONFIG
+
+choice OPENTHREAD_CRYPTO_IMPLEMENTATION
```
Previously, enabling PSA crypto backend didn't depend on using `OPENTHREAD_SECURITY_DEFAULT_CONFIG` which was more flexible because one could enable PSA crypto backend without the default config (e.g. a user may have a custom PSA crypto implementation that uses other configuration scheme). Any reason this needs to be coupled now?

valeriosetti commented on doc/releases/migration-guide-4.4.rst:
```diff
@@ -54,5 +54,11 @@ OpenThread
+  between 2 different crypto providers: legacy Mbed TLS and PSA Crypto API.
+  The former is selected by the newly-added
+  :kconfig:option:`CONFIG_OPENTHREAD_CRYPTO_LEGACY_MBEDTLS`, while the latter
+  from the already existing :kconfig:option:`CONFIG_OPENTHREAD_CRYPTO_PSA`.
```
I'm sorry, I don't understand the requirement here. The doc says that there is the possibility to select the crypto provider so the actionable part for the end user is this one IMO, but I think you are looking for some other information. Can you please clarify or make an example?

valeriosetti committed:
modules: openthread: enable Mbed TLS ECP when using secp256r1 curve

Enabled CONFIG_MBEDTLS_ECP_C also when CONFIG_OPENTHREAD_SRP is enabled as
this is required by CONFIG_MBEDTLS_ECP_DP_SECP256R1_ENABLED.

Signed-off-by: Valerio Setti <vsetti@baylibre.com>

valeriosetti commented on modules/openthread/Kconfig:
```diff
@@ -176,6 +176,7 @@ choice OPENTHREAD_CRYPTO_IMPLEMENTATION
 config OPENTHREAD_CRYPTO_PSA
 	bool "PSA Crypto API"
 	select PSA_CRYPTO
+	select SECURE_STORAGE if !BUILD_WITH_TFM
```
I changed `select` to `imply`. Is this acceptable while we wait for what @tomi-font proposed to be implemented (which seems to belong to a separate PR)?

Damian-Nordic commented on modules/openthread/platform/crypto_psa.c:
```diff
@@ -287,6 +295,160 @@ otError otPlatCryptoHmacSha256Finish(otCryptoContext *aContext, uint8_t *aBuf, s
+
+	operation = (psa_key_derivation_operation_t *) aContext->mContext;
+
+	*operation = psa_key_derivation_operation_init();
```
It's better to `memset` operation to 0. If `operation` is large on a given platform, initializing it with `psa_key_derivation_operation_init()` may needlessly increase the stack usage (we were optimizing it in other functions in the past).

Damian-Nordic commented on modules/openthread/platform/crypto_psa.c:
```diff
@@ -287,6 +295,160 @@ otError otPlatCryptoHmacSha256Finish(otCryptoContext *aContext, uint8_t *aBuf, s
+	otError error = OT_ERROR_NONE;
+	psa_status_t status = PSA_SUCCESS;
+	psa_key_derivation_operation_t *operation = NULL;
+	otCryptoKeyRef keyRef = PSA_KEY_ID_NULL;
```
this file uses `snake_case` like most of Zephyr rather than `camelCase` (except for names that follow OpenThread)

Damian-Nordic commented on modules/openthread/platform/crypto_psa.c:
```diff
@@ -287,6 +295,160 @@ otError otPlatCryptoHmacSha256Finish(otCryptoContext *aContext, uint8_t *aBuf, s
+	uint8_t keyBuffer[kBufferSize];
+
+	if (!checkContext(aContext, sizeof(psa_key_derivation_operation_t)) ||
+	    aInputKey != NULL) {
```
Did you mean `aInputKey == NULL`? Also, should salt be checked for consistency as well?

Damian-Nordic reviewed (commented):
Thanks, the config cleanup looks good. Added remaining comments - the style-related ones are optional by me.

Damian-Nordic reviewed (approved):
LGTM, thanks!

Cristib05 commented:
> > I have added the content of this PR in the openthread border router project, and I tried to run the external commissioner scenario, to obtain Thread network's dataset.
> 
> Thanks for testing this! It's definitely something I wouldn't be able to do on my own because I don't have any board that supports both WiFi and OT.
> 
> > When `Client Key Exchange, Server Key Exchange, Change Cipher Spec, Encrypted Handshake Message is performed", the OTBR responds with 'Handshake failure'. I saw that Client Key Exchange uses EC J-PAKE.
> 
> By any chance do you have any log from the OTBR (with Mbed TLS debug messages enabled)?
> 
> > I've seen some github threads in which is stated that PSA is not fully compliant with ECJPAKE and DTLS1.2, but since I am not very familiar with crypto, I thought it would be better to ask here if you are aware of such thing.
> 
> I will investigate on this with Mbed TLS developers.

I am currently trying to make logging work. I've found something that makes use of mbedtls_ssl_conf_dbg, but I need to understand where the tls context is actually created to make it work. I've only found something related to `heap_init` at the moment, but I'll continue searching. Have you already enabled logging so far?

tomi-font commented on doc/releases/migration-guide-4.4.rst:
```diff
@@ -48,40 +43,37 @@ QSPI
+  :kconfig:option:`CONFIG_PSA_CRYPTO_CLIENT`, but instead selects
+  :kconfig:option:`CONFIG_PSA_CRYPTO`.
+
+* In builds without TF-M, :kconfig:option:`CONFIG_SECURE_STORAGE` is automatically
```
```suggestion
* In builds without TF-M, :kconfig:option:`CONFIG_SECURE_STORAGE` is now automatically
```

Cristib05 commented:
Hi @valeriosetti. I have been debugging more in OpenThread border router application from `samples/net/openthread/border_router`.
At the moment, I have found one issue, related to TREL:
Here is the app call stack:
<img width="294" height="291" alt="image" src="https://github.com/user-attachments/assets/b6f38399-a80d-4479-aba7-c91d1f7ed3cf" />
One issue is in `psa_key_derivation_setup` function. At the moment, the result is `return PSA_ERROR_NOT_SUPPORTED;`
<img width="525" height="157" alt="image" src="https://github.com/user-attachments/assets/34829d81-6153-4b83-8bc6-6fa3c97142b4" />
It seems like `AT_LEAST_ONE_BUILTIN_KDF` has to be defined for an actual call to be made, but I am not sure what flag is required to have this enabled.
<img width="503" height="138" alt="image" src="https://github.com/user-attachments/assets/a67d7b03-71e6-4c29-a609-02c1731784a7" />
The question is what flag I should define not to break the functionality and enable the computation of key?

valeriosetti committed:
modules: openthread: select PSA_WANT deps for KDF algorithms

Select proper PSA_WANT requirements in CONFIG_OPENTHREAD_CRYPTO_PSA_CONFIG
in order to enable KDF algorithms.

Signed-off-by: Valerio Setti <vsetti@baylibre.com>

Cristib05 commented:
> > It seems like AT_LEAST_ONE_BUILTIN_KDF has to be defined for an actual call to be made, but I am not sure what flag is required to have this enabled.
> 
> This is an amazing feedback, thanks! I added a [new commit](https://github.com/zephyrproject-rtos/zephyr/pull/98088/commits/dcf891c5d43f259f6dc5a06a981737d648d52a0a) in the series which hopefully should enable what was missing. Can you please give it another try?

Meanwhile I have managed to have a working joiner-commissioner scenario on OpenThread Border Router sample. I will compare my changes with yours and let you know. Thanks for the new commit!

Cristib05 commented:
> > > It seems like AT_LEAST_ONE_BUILTIN_KDF has to be defined for an actual call to be made, but I am not sure what flag is required to have this enabled.
> > 
> > 
> > This is an amazing feedback, thanks! I added a [new commit](https://github.com/zephyrproject-rtos/zephyr/pull/98088/commits/dcf891c5d43f259f6dc5a06a981737d648d52a0a) in the series which hopefully should enable what was missing. Can you please give it another try?
> 
> Meanwhile I have managed to have a working joiner-commissioner scenario on OpenThread Border Router sample. I will compare my changes with yours and let you know. Thanks for the new commit!

Hi @valeriosetti,
For the joiner-commissioner scenario, I have found that those settings are required:
`select PSA_WANT_ALG_TLS12_PRF if OPENTHREAD_COMMISSIONER_JOINER`
`select PSA_WANT_ALG_TLS12_ECJPAKE_TO_PMS if OPENTHREAD_COMMISSIONER_JOINER`
I also needed to increase `CONFIG_MBEDTLS_HEAP_SIZE` from 10240 to 15360, but I guess this can be further optimized.


Related to your commit from yesterday, 
`select PSA_WANT_ALG_HKDF if OPENTHREAD_COMMISSIONER_JOINER`
`select PSA_WANT_ALG_HKDF_EXTRACT if OPENTHREAD_COMMISSIONER_JOINER`
`select PSA_WANT_ALG_HKDF_EXPAND if OPENTHREAD_COMMISSIONER_JOINER`,
I think that those are not required by `OPENTHREAD_COMMISSIONER_JOINER`. Maybe this should fall under something related to TREL or Border Router, as HKDF-SHA256 is required by TREL.

Related to TREL initialization, I made use of PSA switch proposal from OpenThread, https://github.com/openthread/openthread/pull/11445/commits/3a6573fee80894259dcfe85fc82d9c7e2a7ba520#diff-8d6dc63e1ec9adacf9f4215b293513b17f39a8127ea4b89100db42976adace44,
and for `KeyManager::ComputeTrelKey` to work, a small addition is required in https://github.com/zephyrproject-rtos/zephyr/blob/main/modules/openthread/platform/crypto_psa.c#L97, `OT_CRYPTO_KEY_USAGE_DERIVE` should also be added to the list of supported flags.

valeriosetti commented:
> I think that those are not required by OPENTHREAD_COMMISSIONER_JOINER. Maybe this should fall under something related to TREL or Border Router, as HKDF-SHA256 is required by TREL.

Thanks for the suggestion. I updated the [commit](https://github.com/zephyrproject-rtos/zephyr/pull/98088/commits/7691189d7bd8c205bf4824b05c0cbb8b7704bed4) with the information you provided.

> Related to TREL initialization, I made use of PSA switch proposal from OpenThread, https://github.com/openthread/openthread/commit/3a6573fee80894259dcfe85fc82d9c7e2a7ba520#diff-8d6dc63e1ec9adacf9f4215b293513b17f39a8127ea4b89100db42976adace44, and for KeyManager::ComputeTrelKey to work

I updated the [companion PR](https://github.com/zephyrproject-rtos/openthread/pull/210) to entirely cherry-pick the [commit](https://github.com/openthread/openthread/pull/11445/commits/3a6573fee80894259dcfe85fc82d9c7e2a7ba520) from the PR on the official OT repo. Previously I only took some part of it.

> a small addition is required in https://github.com/zephyrproject-rtos/zephyr/blob/main/modules/openthread/platform/crypto_psa.c#L97, OT_CRYPTO_KEY_USAGE_DERIVE should also be added to the list of supported flags.

Done. Thanks.

valeriosetti committed:
modules: mbedtls: increase heap memory for OpenThread

Increase heap memory when OpenThread is enabled and either "commissioner"
or "joiner" roles are used.

Signed-off-by: Valerio Setti <vsetti@baylibre.com>

tomi-font reviewed (changes requested):
- https://github.com/zephyrproject-rtos/zephyr/pull/98088#discussion_r2548914447
- https://github.com/zephyrproject-rtos/zephyr/pull/98088#discussion_r2548917177

valeriosetti committed:
modules: openthread: move definition of OPENTHREAD_CRYPTO_PSA

Simply move the definition of OPENTHREAD_CRYPTO_PSA Konfig before crypto
configuration related Kconfigs.

Signed-off-by: Valerio Setti <vsetti@baylibre.com>

Cristib05 commented:
> > It seems there are some twister build issues for OpenThread Border Router project. Please allow me some time to understand the issues and potentially solve them. Thanks!
> 
> It seems that failures are related to an assert failure that was [recently added to OT](https://github.com/zephyrproject-rtos/openthread/commit/b3d3b5c3c7de62bdcdd6a5cbabfbd0b545f2e91f#diff-c757d05ba5dcc59650be321cc8ae022be5259c9e91250d947256de26b740dac6R52) and that I picked up in the companion PR which is bumping OT. I will look into how to fix this.

I guess my PR https://github.com/zephyrproject-rtos/zephyr/pull/102649 should solve them.
